dre-service/src/main/java/io/cdap/re/RulesDB.java

```diff
@@ -162,19 +169,23 @@ void createRule(RuleRequest rule) throws RuleAlreadyExistsException, IOException
     rulesTable.upsert(ruleFields);
   }
 
-  void updateRule(String id, RuleRequest rule) throws RuleNotFoundException, IOException {
+  void updateRule(NamespaceId namespaceId, String id, RuleRequest rule) throws RuleNotFoundException, IOException {
```

Comment:
should introduce a class that represents an id instead of having different arguments. See https://github.com/data-integrations/wrangler/blob/develop/wrangler-proto/src/main/java/io/cdap/wrangler/proto/NamespacedId.java for an example.

The namespace generation is used to handle the scenario where somebody deletes a namespace, then creates a new namespace with the same name. In that case, rules that existed before the namespace was deleted should no longer be visible.

Reply:
What is `NamespacedId` used for? I looked at how the stores are implemented in Wrangler and they all use the name and generation from a `Namespace` object, not a `NamespacedId`.

Reply:
the key is 3 things, namespace, generation, and "id". These 3 things uniquely identify an object.

Reply:
And what is the "id"? Because, I understand that a Namespace has a name and generation, but what is the id? The `String id` parameter in this method is the id of the rule, not the Namespace.

Reply:
The "id" is what used to be the id of the rule. "id" is a poor name for it now though, since it no longer uniquely identifies the object. It's now uniquely identified by namespace, generation, and "id", hence the use of the NamespacedId class as an identifier.